### PR TITLE
Mas d31 i433re2capture

### PR DESCRIPTION
**Pull request by martinsumner**
Allow for evaluation and filtering of 2i terms.

Eval: Have a language of functions whereby, a map of Key/Value pairs can be extracted by applying those functions in a pipeline.

Filter: Have a language of functions and operators to apply a logic expression to determine if a term matches a filter, based on the entries in the map output from the eval pipeline.

The aim is to have an alternative to regular expression filtering in Riak.  Sometimes the auto-generation of regular expressions become complex, and somethings (such as date ranges) and potentially impossible.  The aim of the filter expression is to do this in a more natural language.

### CODE REVIEW

File: src/leveled_codec.erl [hunk 1]
```diff
@@ -112,10 +112,17 @@
         {index_specs(), infinity|integer()}. % {KeyChanges, TTL}
 -type maybe_lookup() ::
         lookup|no_lookup.
+-type actual_regex() ::
+        reference()|{re_pattern, term(), term(), term(), term()}.
```
Comment (ThomasArts) [inline src/leveled_codec.erl:116]: is there a reason not to use the opaque type `re:mp()` instead of `re_patten...`?

File: src/leveled_eval.erl [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+apply_eval({eval, Eval}, Term, Key, AttrMap) ->
+    apply_eval(Eval, Term, Key, AttrMap);
+apply_eval({'PIPE', Eval1, 'INTO', Eval2}, Term, Key, AttrMap) ->
+    apply_eval(Eval2, Term, Key, apply_eval(Eval1, Term, Key, AttrMap));
+apply_eval({
+        delim, {identifier, _, InKey}, {string, _, Delim}, ExpKeys},
+        Term, Key, AttrMap) ->
+    TermToSplit = term_to_process(InKey, Term, Key, AttrMap),
+    CptTerms = string:split(TermToSplit, Delim, all),
+    L = min(length(CptTerms), length(ExpKeys)),
```
Comment (ThomasArts) [inline src/leveled_eval.erl:39]: in OTP 26 there is now a `How` parameter for zip. Using `trim` would reduce this code. But only introduced for OTP26.

**Review by ThomasArts — APPROVED** [review]

File: src/leveled_codec.erl (hunk 1 above)
Comment (martinsumner) [reply src/leveled_codec.erl:116]: [Not exported until > OTP26](https://github.com/erlang/otp/commit/eb4a9c1e45bccf8d5463ec69198db41149eda823)

**Comment by martinsumner** [comment]
Replaced with https://github.com/martinsumner/leveled/pull/440

**Pull request closed without merging**